lang/csharp/src/apache/main/IO/BinaryDecoder.notnetstandard2.0.cs

```diff
@@ -0,0 +1,100 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Text;
+
+namespace Avro.IO
+{
+    /// <content>
+    /// Contains the netstandard2.1 and netcoreapp2.1 specific functionality for BinaryDecoder.
+    /// </content>
+    public partial class BinaryDecoder
+    {
+        private const int StackallocThreshold = 256;
+
+        /// <summary>
+        /// A float is written as 4 bytes.
+        /// The float is converted into a 32-bit integer using a method equivalent to
+        /// Java's floatToIntBits and then encoded in little-endian format.
+        /// </summary>
+        /// <returns></returns>
+        public float ReadFloat()
+        {
+            Span<byte> buffer = stackalloc byte[4];
+            Read(buffer);
+
+            return BitConverter.Int32BitsToSingle(BinaryPrimitives.ReadInt32LittleEndian(buffer));
+        }
+
+        /// <summary>
+        /// A double is written as 8 bytes.
+        /// The double is converted into a 64-bit integer using a method equivalent to
+        /// Java's doubleToLongBits and then encoded in little-endian format.
+        /// </summary>
+        /// <returns>A double value.</returns>
+        public double ReadDouble()
+        {
+            Span<byte> buffer = stackalloc byte[8];
+            Read(buffer);
+
+            return BitConverter.Int64BitsToDouble(BinaryPrimitives.ReadInt64LittleEndian(buffer));
+        }
+
+        /// <summary>
+        /// Reads a string written by <see cref="BinaryEncoder.WriteString(string)"/>.
+        /// </summary>
+        /// <returns>String read from the stream.</returns>
+        public string ReadString()
+        {
+            byte[] bufferArray = null;
+
+            int length = ReadInt();
+            Span<byte> buffer = length <= StackallocThreshold ?
+                stackalloc byte[length] :
+                (bufferArray = ArrayPool<byte>.Shared.Rent(length));
+
+            Read(buffer);
+
+            string result = Encoding.UTF8.GetString(buffer);
+
+            if (bufferArray != null)
+            {
+                ArrayPool<byte>.Shared.Return(bufferArray);
+            }
+
+            return result;
+        }
+
+        private void Read(byte[] buffer, int start, int len)
+        {
+            Read(buffer.AsSpan(start, len));
+        }
+
+        private void Read(Span<byte> buffer)
+        {
+            while (!buffer.IsEmpty)
+            {
+                int n = stream.Read(buffer);
+                if (n <= 0) throw new AvroException("End of stream reached");
+                buffer = buffer.Slice(n);
```

Comment:
Do you happen to know if any of our unit tests are currently exercising this code? If not, could we change them (I guess through adding TargetFrameworks?) to do so?

If I'm following this logic correctly, `buffer` will always be empty when this method returns. Am I misreading this? I'll try to step into this with the debugger tomorrow when I've got my normal dev environment.

Reply:
> Do you happen to know if any of our unit tests are currently exercising this code?

Yes, they are. The tests run for both .NET Framework and on .NET Core:

https://github.com/apache/avro/blob/9bc2c62c6a82dbce260d4cc0277f1d0276063468/lang/csharp/src/apache/test/Avro.test.csproj#L21

> If I'm following this logic correctly, buffer will always be empty when this method returns. Am I misreading this?

You are following it correctly and not misreading it. The `Read` method expects to read `buffer.Length` bytes from the stream, and it loops reading from the stream until the whole buffer is filled.  (Note that the callers `Span` is not updated, so the caller is still pointing to the beginning of the buffer upon return.)

Reply:
Ah, that makes sense. I clearly haven't studied up on `Span<T>` enough yet. Thank you for explaining this!